7066: Dataset Designer Server Side API updates

RosalineP opened the pull request:
Some notes:
- We don't currently check to make sure that the dataset name is under a maximum amount of characters—do we want to add that?
- Current code allows for a keyPropertyName that is the name of a column that is not numeric or string in type. Should this be changed? (Edit: This constraint to the type of the third key column only applies if the third key column is managed.)
- If you change the name of a dataset in the options but not the domainDesign, it saves and changes the dataset name. Is this expected? The save does not go through if you change the domainDesign name but not the options name.
- Category issue I noted

labkey-ankurj committed:
Item 6642: Starter files DatasetDesigner react app

labkey-ankurj committed:
Merge branch 'develop' into fb_datasetDesigner

RosalineP committed:
Merge remote-tracking branch 'remotes/origin/fb_datasetDesigner' into fb_datasetDesignerStory3_7066

RosalineP committed:
Merge branch 'develop' into fb_datasetDesignerStory3_7066

RosalineP committed:
Work on moving functionality to DatasetDomainKind

RosalineP committed:
Merge branch 'develop' into fb_datasetDesignerStory3_7066

RosalineP committed:
Merge branch 'develop' into fb_datasetDesignerStory3_7066

RosalineP committed:
Merge branch 'develop' into fb_datasetDesignerStory3_7066

RosalineP committed:
Progress on managers, DDK, and DDKP

RosalineP committed:
Fix typo

RosalineP committed:
Merge branch 'develop' into fb_datasetDesignerStory3_7066

RosalineP committed:
Gets callable; update written and to be tested

RosalineP committed:
Merge branch 'develop' into fb_datasetDesignerStory3_7066

RosalineP committed:
Fix bugs in create and get

RosalineP committed:
Debugging update

labkey-ankurj committed:
Merge branch 'develop' into fb_datasetDesignerStory3_7066

RosalineP committed:
Merge branch 'develop' into fb_datasetDesignerStory3_7066

RosalineP committed:
Save, create, and get actions working. Continues work on validation

labkey-ankurj committed:
Item 7066: partial wrapper around DatasetDesigner

labkey-ankurj committed:
Merge remote-tracking branch 'origin/fb_datasetDesignerStory3_7066' into fb_datasetDesignerStory3_7066

RosalineP committed:
Adds read-only properties to getDomainKindProperties. Addresses areas flagged by me

RosalineP committed:
Merge remote-tracking branch 'origin/fb_datasetDesignerStory3_7066' into fb_datasetDesignerStory3_7066

cnathe committed:
Study module package.json updates to match version and other minor updates made to related platform modules

RosalineP committed:
Applies feedback from question-review round

RosalineP committed:
Merge remote-tracking branch 'origin/fb_datasetDesignerStory3_7066' into fb_datasetDesignerStory3_7066

labkey-ankurj committed:
Item 7066: partially working dataset designer wrapper to view existing properties of dataset

labkey-ankurj committed:
Item 7066: partially working save dataset designer

RosalineP committed:
Handles more edge cases for create and update

RosalineP committed:
Removes temp notes to myself

labkey-jeckels commented:
@RosalineP 
1. What's the behavior if you pass in a very long dataset name? study.dataset.name is 200 characters, so that's what you'd want to be sure to exceed. It might give you a decent enough error message already.

2. Yes, restricting to just integer and string fields makes sense.

3. It's not great, but what you're describing probably matches what happens if you edit a dataset's name through the GWT designer. I'm not sure there's a lot of value in having duplicate copies of the value in the JSON, only one of which matters when you save.

RosalineP committed:
Tweaks exception catching. Changes isDemographicData to demographicData for consistency

RosalineP committed:
Fixes category issue

cnathe commented:
> @RosalineP
> 
> 1. What's the behavior if you pass in a very long dataset name? study.dataset.name is 200 characters, so that's what you'd want to be sure to exceed. It might give you a decent enough error message already.
> 2. Yes, restricting to just integer and string fields makes sense.
> 3. It's not great, but what you're describing probably matches what happens if you edit a dataset's name through the GWT designer. I'm not sure there's a lot of value in having duplicate copies of the value in the JSON, only one of which matters when you save.

For #2, don't we want to allow any of the fields to be used for the additional key (i.e. keyPropertyName) for the non-managed case? Doesn't the type restriction only come into play when the additional key is also selected as a "managed" 3rd key (i.e. one that LK server will populate for you)?

cnathe commented on study/src/org/labkey/study/controllers/StudyController.java:
```diff
@@ -463,13 +458,8 @@ public boolean handlePost(ImportTypeForm form, BindException derrors)
-                return true;
-            }
-            return false;
+            ((DatasetDefinition)_def).provisionTable();
+            return true;
```
@RosalineP @labkey-ankurj Either of you know why this change was made? Seems unrelated to this PR.

RosalineP replied:
It looks like this got pulled in when I initially pulled in Ankur's react page setup changes on fb_datasetDesigner: https://github.com/LabKey/platform/compare/fb_datasetDesigner
@labkey-ankurj Do we still want it?

labkey-ankurj replied:
@cnathe I see these are not recent changes and not sure why they are getting displayed in this diff.
![Screen Shot 2020-04-17 at 7 43 54 AM](https://user-images.githubusercontent.com/39965420/79581425-42e92900-807f-11ea-94cf-8fb996c9f953.png)

cnathe committed:
Item 7066: misc PR cleanup during code review
- update README.md with the module description from module.properties file
- Clean up imports in StudyController and StudyService

labkey-jeckels commented:
Yes, thanks @cnathe . I was only thinking about the managed case. We shouldn't restrict to int and string for unmanaged keys.

RosalineP commented:
Yup, that's right, I should have clarified that—I was only talking about the managed third key case

cnathe commented on study/src/org/labkey/study/model/DatasetDomainKind.java:
```diff
@@ -314,32 +330,64 @@ public boolean canCreateDefinition(User user, Container container)
+        Dataset ds = getDatasetDefinition(domain.getDomainURI());
+        DatasetDomainKindProperties datasetProperties = new DatasetDomainKindProperties(ds);
+
+        if (container.isProject() && StudyService.get().getStudy(container).isDataspaceStudy())
+        {
+            datasetProperties.setDefinitionIsShared(true);
+            if (StudyService.get().getStudy(container).getShareVisitDefinitions())
+                datasetProperties.setVisitMapShared(true);
+        }
```
@RosalineP This part needs to be in the DatasetDomainKindProperties constructor itself instead of here. That way both code paths that use the `new DatasetDomainKindProperties(ds)` will get these set. 

The code to put into the DatasetDomainKindProperties constructor should be this, I believe:
```
Study study = StudyService.get().getStudy(ds.getContainer());
        if (ds.getContainer().isProject() && study.isDataspaceStudy())
        {
            _definitionIsShared = study.getShareDatasetDefinitions();
            _visitMapShared = study.getShareVisitDefinitions();
        }
```

cnathe replied:
actually since that is based off of the study itself and not specific to the dataset, we'll need that in the default constructor case as well. So this should go in the `public DatasetDomainKindProperties() {}` constructor and then we'll need to call it with the container as a param for that constructor.

RosalineP replied:
Oh, that makes sense. Have added it as suggested to both constructors.

labkey-ankurj replied:
Should this be added in the constructor as well?
![Screen Shot 2020-04-16 at 4 35 49 PM](https://user-images.githubusercontent.com/39965420/79516534-8b153680-8000-11ea-8e61-c9934298c2a5.png)

cnathe commented on study/src/org/labkey/study/model/DatasetDomainKind.java:
```diff
@@ -314,32 +330,64 @@ public boolean canCreateDefinition(User user, Container container)
+            throw new IllegalArgumentException("\"Additional Key Column name \"" + keyPropertyName +"\" must be the name of a column");
+
+        if (demographics && (isManagedField || keyPropertyName != null))
+            throw new IllegalArgumentException("There can not be an Additional Key Column if the dataset is Demographic Data");
```
@RosalineP as discussed, can some of these validation checks that apply to both the create and update case be factored out to a separate method which both createDomain() and updateDomain() can call?

cnathe commented on study/src/org/labkey/study/model/DatasetDomainKind.java:
```diff
@@ -435,6 +476,212 @@ else if (categoryId != null)
         }
     }
 
+    private ValidationException validateDatasetProperties(DatasetDefinition def, ValidationException exception, Container container, User user,
```
@RosalineP from the comment above, maybe this method is the one that has the validation checks that are common between the create and update case?

RosalineP replied:
Because this validateDatasetProperties() function has some stuff that can't be called by the create case, I've reverted its name to checkCanUpdate() and created a different, shared validateDatasetProperties() function

RosalineP replied:
I ended up having a couple of questions about reusing the existing validation inside of this function across create and update. I have a temporary implementation right now that reverts this validateDatasetProperties() to the original name of 'checkCanUpdate' and creates a different validateDatasetProperties() function that shares the common validation checks—I'll ping you about those questions!

cnathe reviewed (commented):
Adding some initial comments, I will re-review the DatasetDomainKind changes in more detail after Rosaline is done with the last part she is currently working on.

cnathe committed:
Item 7066: misc PR cleanup during code review
- DatasetDomainKindProperties constructor reorder for code readability

cnathe committed:
Merge branch 'develop' into fb_datasetDesignerStory3_7066

cnathe commented on study/src/client/DatasetDesigner/DatasetDesigner.tsx:
```diff
@@ -0,0 +1,154 @@
+                <DatasetDesignerPanels
+                    initModel={model}
+                    onCancel={this.onCancel}
+                    showDataSpace={false}
```
@labkey-ankurj You may have already seen this, but looks like the relevant value for this is being returned with the dataset model: `definitionIsShared`

labkey-ankurj replied:
Yes, I have addressed this and it's in my local set of changes.

labkey-ankurj committed:
Item 7066: set definition shared changes

labkey-ankurj committed:
Item 7066: handle window before unload

RosalineP committed:
Addresses feedback: modularizes exception checking, favors throwing IllegalArgumentExceptions, and moves initialization of DatasetDomainKindProperties

RosalineP committed:
Undoes breaking changes from last commit. Work ongoing on verifying all validation cases

RosalineP committed:
Fixes validation exceptions

RosalineP committed:
One last tweak

cnathe committed:
Item 7066: remove redundant code for setting DatasetDomainKindProperties for study properties

cnathe committed:
Item 7066: remove unused imports (missed from last commit)

labkey-ankurj committed:
pulling latest version of shared components

cnathe committed:
Item 7066: add punctuation to error messages

cnathe committed:
Item 7066: add missing @Override annotations

cnathe committed:
Item 7066: DatasetDomainKind code rearrangement during code review

cnathe commented on study/src/org/labkey/study/model/DatasetDomainKind.java:
```diff
@@ -356,20 +374,12 @@ else if (!timepointType.isVisitBased() && getKindName().equals(VisitDatasetDomai
             else
                 throw new IllegalArgumentException("Unable to find a category named : " + categoryName + " in this folder.");
```
Looks like the createDomain case for a dataset does not currently support the "create a new category by name and use it for this dataset". The updateDomain case supports this via its call to `StudyManager.getInstance().updateDatasetDefinition`, which ends up calling `ensureViewCategory()` in the StudyManager. We should sync up on how to handle this for the create dataset scenario.

RosalineP replied:
As of last commit (927c8e2) creating a new category with a new dataset is working

cnathe commented on study/src/org/labkey/study/model/DatasetDomainKind.java:
```diff
@@ -435,6 +443,231 @@ else if (categoryId != null)
+        try
+        {
+            // Update-case specific validation
+            if ( datasetProperties.isDemographicData() && !datasetProperties.isDemographicData() && !StudyManager.getInstance().isDataUniquePerParticipant(def))
```
Looks like both of the first two parts of this if statement are using `datasetProperties`, I believe the second one should be using the `DatasetDefinition def` param:
```suggestion
            if ( datasetProperties.isDemographicData() && !def.isDemographicData() && !StudyManager.getInstance().isDataUniquePerParticipant(def))
```

cnathe replied:
Also, I think we should move this validation check to the `checkCanUpdate()` method so that it lives with the other update dataset specific validation checks.

cnathe commented on study/src/org/labkey/study/model/DatasetDomainKind.java:
```diff
@@ -435,6 +443,231 @@ else if (categoryId != null)
+
+            // PR Note: is it desired that we check for this at all?
+//            if (null != datasetProperties.getCategory() && null == ViewCategoryManager.getInstance().getCategory(container, datasetProperties.getCategory()))
+//                throw new IllegalArgumentException("Unable to find a category named : " + datasetProperties.getCategory() + " in this folder.");
```
See previous comment in this code review. For the update case, we already have the proper checks happening for the category (i.e. to use an existing one or create a new one) via the StudyManager code path. So these commented out lines can be removed.

cnathe commented on study/src/org/labkey/study/model/DatasetDomainKind.java:
```diff
@@ -435,6 +443,231 @@ else if (categoryId != null)
+            {
+                datasetProperties.setKeyPropertyName(null);
+                datasetProperties.setKeyPropertyManaged(false);
+            }
```
It looks like the `validateDatasetProperties` already checks for this case, so I think this if statement is basically a no-op. That is, if the property is set as demographic we would have already thrown an exception if the keyProperty name wasn't null or the managed flag was true. I think this can be removed.

cnathe commented on study/src/org/labkey/study/model/DatasetDomainKind.java:
```diff
@@ -435,6 +443,231 @@ else if (categoryId != null)
+
+            /* IGNORE illegal shareDataset values */
+            if (!study.getShareVisitDefinitions())
+                datasetProperties.setDataSharing("NONE");
```
The comment here says to ignore the illegal argument for sharedDataset, I think we should instead throw an exception for this case instead of just changing it for the user/caller. I suggest moving this to the `validateDatasetProperties` method and have the check throw an error if the study does not allow for shared visit definitions and the dataSharing property is anything but NONE.

cnathe commented on study/src/org/labkey/study/model/DatasetDomainKind.java:
```diff
@@ -435,6 +443,231 @@ else if (categoryId != null)
+                exception.addGlobalError(errorMsg);
+            return exception;
+        }
+        catch (RuntimeSQLException e)
+        {
+            return exception.addGlobalError("Additional key column must have unique values.");
+        }
```
I don't see where this is throw from, maybe it can be deleted?

RosalineP replied:
Yeah, I agree and have deleted it—it looks as though this is trying to gate for is the third key column is changed to a column that has duplicates and violates the participant + visit + key uniqueness constraint? Does that sound to you like something we should gate for in validateDatasetProperties()?

cnathe replied:
Ah ok yes, that is the scenario. So if you have a dataset configured and saved which already has data in it, then you go into the designer and set an additional key column for that dataset but the data in the table violates that constraint (i.e. the ptid/timepoint/key are unique), then we'd expect an error. Maybe this catch is the right place for that.

RosalineP replied:
Cool, I've put it back for now

cnathe commented on study/src/org/labkey/study/model/DatasetDomainKind.java:
```diff
@@ -435,6 +443,231 @@ else if (categoryId != null)
+        }
+        catch (Exception x)
+        {
+            throw UnexpectedException.wrap(x);
```
The createDomain() case uses `throw new RuntimeException(e);` for this general catch. We should make them consistent.

cnathe commented on study/src/org/labkey/study/model/DatasetDomainKind.java:
```diff
@@ -435,6 +443,231 @@ else if (categoryId != null)
+
+        checkCanUpdate(def, container, user, original, update);
+
+        // Remove any fields that are duplicates of the default dataset fields (e.g. participantid, etc).
+        List<? extends GWTPropertyDescriptor> updatedProps = update.getFields();
+        for (Iterator<? extends GWTPropertyDescriptor> iter = updatedProps.iterator(); iter.hasNext();)
+        {
+            GWTPropertyDescriptor prop = iter.next();
+            if (DatasetDefinition.isDefaultFieldName(prop.getName(), study))
+                iter.remove();
+            else if (DatasetDomainKind.DATE.equalsIgnoreCase(prop.getName()))
+                prop.setRangeURI(PropertyType.DATE_TIME.getTypeUri());
+        }
```
This code confuses me. It looks like this "remove duplicate default fields" only happens in the create dataset case. I wonder if we either want to have it happen for both create and update cases or if we should just remove it and let the "reserved field name" validation check handle this case?

RosalineP replied:
I tested this through and it looks like the reserved field name validation check does catch for if a field name is a duplicate of an existing field name—and it also provides a nice error message rather than silently removing the offending column, which is good. I've removed this bit of code

cnathe committed:
Item 7066: DatasetDomainKind code rearrangement during code review (part 2)

cnathe committed:
Merge branch 'develop' into fb_datasetDesignerStory3_7066

cnathe commented on study/src/client/DatasetDesigner/DatasetDesigner.tsx:
```diff
@@ -0,0 +1,158 @@
+    isLoadingModel: boolean,
+    message?: string,
+    dirty: boolean,
+    hasDatasetDesignPermission?: boolean,
```
@labkey-ankurj this doesn't look to be used at all in this component/file.

cnathe commented on study/src/client/DatasetDesigner/DatasetDesigner.tsx:
```diff
@@ -0,0 +1,158 @@
+                    initModel={model}
+                    onCancel={this.onCancel}
+                    useTheme={true}
+                    onComplete={this.onComplete}
```
@labkey-ankurj  Two things missing here `successBsStyle={'primary'}` and the `onChange` prop/listener. See list/Designer.tsx.

labkey-ankurj replied:
added

RosalineP committed:
Applying feedback

RosalineP committed:
Adds category creation for new dataset

labkey-ankurj committed:
pulling in new shared components version

labkey-ankurj committed:
upgrading ui version

cnathe committed:
Update to the latest @labkey/components alpha package version

cnathe committed:
Add study/node_modules to .gitignore at platform

cnathe committed:
Remove unused "studyId" prop from the study moduleContext

RosalineP committed:
Fix for description, label, and tag

labkey-ankurj committed:
Pulling in updated ui-components version

RosalineP committed:
Fix for Visit Date Column

RosalineP committed:
Merge remote-tracking branch 'origin/fb_datasetDesignerStory3_7066' into fb_datasetDesignerStory3_7066

cnathe committed:
AssayPublishManager.createAssayDataset refactor from previous commit

cnathe committed:
Revert change in DefineDatasetTypeAction.handlePost

RosalineP committed:
Support useTimeKeyField on DatasetDomainKindProperties

RosalineP committed:
Merge remote-tracking branch 'origin/fb_datasetDesignerStory3_7066' into fb_datasetDesignerStory3_7066

labkey-ankurj committed:
Disable phi level for additional key field

labkey-ankurj committed:
code review changes on platform app

cnathe committed:
Merge branch 'develop' into fb_datasetDesignerStory3_7066

# Conflicts:
#	assay/package-lock.json
#	assay/package.json
#	core/package-lock.json
#	core/package.json
#	experiment/package-lock.json
#	experiment/package.json
#	list/package-lock.json
#	list/package.json
#	query/package-lock.json
#	query/package.json

cnathe committed:
Update to the latest @labkey/components alpha package version

cnathe committed:
Dataset create case needs to handle the showByDefault and label properties

cnathe committed:
Update to the latest @labkey/components alpha package version

labkey-ankurj committed:
column mapping changes for datetime vs dateTime

labkey-ankurj committed:
pulling in changes for additional key field and visit date field changes

cnathe commented on study/package.json:
```diff
@@ -0,0 +1,44 @@
+    }
+  },
+  "dependencies": {
+    "@labkey/components": "0.52.1-fb-datasetDesigner.6"
```
@labkey-ankurj I have been bumping the other @labkey/components package usages in the platform repo along with this one so that we can get TeamCity to run with all the related modules (i.e. query, core, list, experiment, assay, study).

labkey-ankurj committed:
pulling in shared components designer version in platform apps

cnathe committed:
Update to the latest @labkey/components alpha package version

RosalineP committed:
For create case, allows empty label in API call--populates label with name given

cnathe committed:
Update to the latest @labkey/components alpha package version

cnathe committed:
Add dataset designer message and link about datasets created from assay copy-to-study

cnathe committed:
Add entityId to the DatasetDomainKindProperties so it can be used to determine if a dataset is new or existing

cnathe committed:
Fix for dataset validation check on managed field setting for additional key

labkey-ankurj committed:
pulling in latest alpha version

labkey-ankurj committed:
pulling in latest alpha version

labkey-ankurj committed:
pulling in latest alpha version

cnathe committed:
Support for dataSharing prop in createAssayDataset via DatasetDomainKind.createDomain() call

cnathe committed:
Update to the latest @labkey/components alpha package version

labkey-ankurj committed:
Merge branch 'develop' into fb_datasetDesignerStory3_7066

labkey-ankurj committed:
pulling in latest alpha version

cnathe committed:
Automated test fix to account for dataset domain update without properties / options

RosalineP committed:
Updates member variables to be private, and provides backwards compatibility for 'demographics' option

RosalineP committed:
Adds comment about backwards compatibility and removes related setter

labkey-ankurj committed:
merge develop into branch 'fb_datasetDesignerStory3_7066'

labkey-ankurj committed:
pull beta version

cnathe committed:
Add in a few missing validation checks for DatasetDomainKind handling of useTimeKeyField

labkey-chrisj committed:
Item #6799 Dataset Designer UX test migration

RosalineP committed:
Merge branch 'develop' into fb_datasetDesignerStory3_7066

cnathe committed:
Revert "Item #6799 Dataset Designer UX test migration"

This reverts commit 46e8b1b826e4534d62bf939854817b2f23407bbc.

cnathe committed:
Merge branch 'develop' into fb_datasetDesignerStory3_7066
